packages/docs/docs/guides/performance.md

```diff
@@ -0,0 +1,177 @@
+
+By default, AI.JSX streams responses for you.
+
+### From UI
+
+```tsx file="app.tsx"
+// React component
+<ResultContainer title={`AI comes up with a recipe for ${query}`}>
+  <AI.jsx>
+    {/* AI.JSX components */}
+    <ChatCompletion temperature={1}>
+      <UserMessage>Write me a poem about {query}</UserMessage>
+    </ChatCompletion>
+  </AI.jsx>
+</ResultContainer>
+```
+
+This will call the chat model and automatically stream the results into the DOM.
+
+### From the API
+
+```tsx
+function MyComponent() {
+  return (
+    <ChatCompletion temperature={1}>
+      <UserMessage>Write me a poem about dogs</UserMessage>
+    </ChatCompletion>
+  );
+}
+
+/** Get the partial result as it's streamed */
+function handleIntermediateResult() {}
+
+const finalResult = await AI.createRenderContext().render(<MyComponent />, {
+  map: handleIntermediateResult,
+});
+```
+
+For more detail, see [Intermediate Results](./rules-of-jsx.md#intermediate-results).
+
+## Minimize how long the output needs to be
+
+A model generation takes linearly more time as the output length increases, so the shorter your output can be, the faster it'll be. However, this inherently means the model is spending less time "thinking" about your result, [which could degrade accuracy](./brand-new.md#thinking-out-loud). You'll need to balance these tradeoffs.
+
+For example, you may have a scenario where you instruct the model to give you output in a structured format, like JSON. If you can come up with a simpler JSON format that takes fewer characters, the model will produce it faster.
+
+## Avoid waterfalls / roundtrips
+
+Just like with UI engineering, waterfalls can hurt your performance. Sometimes they're unavoidable, but be mindful when introducing them.
+
+For instance:
+
+```tsx
+function CharacterGenerator() {
+  return <ChatCompletion>
+    <UserMessage>Write me a profile for a fantasy character</UserMessage>
+  </ChatCompletion>
+}
+
+function ToJson({children}: {AI.Node}) {
+  return <ChatCompletion>
+    <UserMessage>Convert this to JSON: {children}</UserMessage>
+  </ChatCompletion>
+}
+
+function App() {
+  return <ToJson><CharacterGenerator /></ToJson>
+}
+```
+
+This is a waterfall, because the `CharacterGenerator` call needs to complete before the `ToJson` call can start.
+
+Instead, we could combine them into a single model call:
+
+```tsx
+function CharacterGenerator() {
+  return (
+    <ChatCompletion>
+      <SystemMessage>Respond in JSON.</SystemMessage>
+      <UserMessage>Write me a profile for a fantasy character</UserMessage>
+    </ChatCompletion>
+  );
+}
+```
+
+If you can get the model to do what you want in a single shot, that'll be more performant. However, asking the model to do more at once decreases reliability. It's more split your workload into simple, discrete tasks for the model. So, there are tradeoffs to balance here. You want the task size to be as complicated as the model can reliably do in a single pass, but no more complicated.
```

Comment:
The "It's more" sentence seems like it's missing something.

Reply:
whoops. Good catch.